Make cmake work better

Justasic opened the pull request:
I have added some stuff to make cmake work a bit better between not just linux distributions but also on windows if you decide to port the project to windows.

The following fixes were made:
- CMake now tries to find boost, curl, and the required threading dependencies itself (curl was actually a missing dependency)
- CMake now downloads the latest version of nlohmann/json automatically on configure so the local version can be removed from the repository.
- Disallowed source directory builds (because those are messy) and instead encourage you make a build directory
- Made CMake determine what C++ flags we use for setting the C++11 standard, this guarantees that cmake always uses C++11 or fails to configure.
- I was forced to change the CMake required version to 3.1.3 because of the C++ standard stuff above. Most systems that support C++11 also have cmake 3.1.3 or higher

Commit:
[CMake] Make CMake more platform agnostic

Make cmake figure out stuff instead of us figuring it out for cmake,
     allowing us to compile on other platforms with more ease.
Added curl as a requirement in CMake.
Fixed finding threading for both C++11 and boost in cmake.
Download nlohmann's json library instead of keeping a local copy.

NotUnlikeTheWaves commented on CMakeLists.txt:
```diff
@@ -1,22 +1,58 @@
+# Download JSON.hpp
+if (NOT EXISTS ${CMAKE_BINARY_DIR}/external/json/json.hpp)
+    message(STATUS "Downloading nlohmann's json.hpp library...")
+    file(DOWNLOAD https://raw.githubusercontent.com/nlohmann/json/develop/src/json.hpp ${CMAKE_BINARY_DIR}/external/json/json.hpp SHOW_PROGRESS)
```
Perhaps it's an option to get it from the master branch?

NotUnlikeTheWaves commented on CMakeLists.txt:
```diff
@@ -1,22 +1,58 @@
+find_package(Threads REQUIRED)
 
 # Disable tests for JSON
 set(BuildTests OFF CACHE INTERNAL "" FORCE)
```
not review related but I this can be deleted if I'm correct, since nlohmann/json is only used as its raw header file instead of full repo w/ CMakeLists.txt.

NotUnlikeTheWaves reviewed (commented):
woah, cool stuff.

Justasic commented:
I found some new CMake code that lets me clone the whole nlohmann repo via ExternalProject as part of the build process and not build their cmake file. This will work better instead of the weird file download. I didn't do that before because ExternalProject wants to build nlohmann's tests and I didn't know how to disable it.

hiddos commented:
originally I moved away from cloning nlohmann because it takes up a 200mb download (and disk space), most of it twats

Justasic commented:
I'll see if I can make it so it only does a clone of one depth of history. If it's still too much I can just make it download like it does now. I misunderstood the original comment, I can add an option to download from the master branch (or any branch I guess for that matter)

hiddos commented:
I meant it more that the master branch I expect to be more stable than the develop branch. If a shallow clone doesn't result in much disk usage I'm fine with that, too

Commit:
[websocketpp] Project has bug fixes for newer versions of OpenSSL, update.

Websocketpp has updated and brought some bug fixes on their develop
branch which fixed a lot of issues with systems that had newer
openssl installations. (Fixes builds on my arch system)
